core/driver/sources/src/main/scala/org/virtuslab/ideprobe/WorkspaceProvider.scala

```diff
@@ -114,11 +114,10 @@ object WorkspaceTemplate {
 
   case class FromGit(repository: String, ref: Option[String]) extends WorkspaceTemplate {
     override def setupIn(workspace: Path): Unit = {
-      val cloned = Shell.run("git", "clone", repository, workspace.toString)
-      if (cloned.exitCode != 0) throw new IllegalStateException(s"Could not clone git $repository")
-      ref.foreach { ref =>
-        val checkout = Shell.run(in = workspace, "git", "checkout", ref)
-        if (checkout.exitCode != 0) throw new IllegalStateException(s"Could not checkout $ref in $repository")
+      import org.virtuslab.ideprobe.dependencies.git.GitHandler._
+      val git = repository.clone(workspace)
+      ref.foreach{
```

Comment:
Maybe the api for git can be cleaned up a bit without many implicit conversions:
let's have a factory method to create git repo by cloning it to some location and such repo will have the necessary methods.
```
val repo = GitRepository.clone(repository, workspace)
ref.foreach(repo.checkout)
```

The GitRepository already exists in dependencies though and it looks similar to what I suggested. 
I guess it could stay as is, but it would be ideal to refactor all git interactions in this repo to rely on one common api.

Reply:
Well, some proper refactoring would be to:
have some (new) GitRepository class with git operations like checkout, push, etc. (whatever we need)
This class can have static factory methods like GitRepository.clone(from, branch?), GitRepository.existing(path) (and maybe some useful method cloneAndCheckout.
Only this class would contain any git logic and would use jGit under the hood, report progress etc.
Any other classes that represent path to git repository and a branch that we use to setup pants or workspace would be plain data classes. These data classes could be converted to git repository, either explicitly: GitRepository.cloneAndCheckout(repo.path, repo.ref) or with some helper methods.

Reply:
@lukaszwawrzyk can we mark it as solved?